src/Mod/Part/Gui/ViewProviderGridExtension.cpp

```diff
@@ -0,0 +1,548 @@
+        return;
+
+    Gui::View3DInventorViewer* viewer = view->getViewer();
+
+    bool cameraDimensionsChanged = checkCameraZoomChange(viewer);
+
+    bool cameraCenterMoved = checkCameraTranslationChange(viewer);
+
+    bool gridNeedUpdating = cameraDimensionsChanged || cameraCenterMoved;
+
+    if (!gridNeedUpdating && cameraUpdate)
+        return;
+
+    Gui::coinRemoveAllChildren(GridRoot);
+
+    computeGridSize(viewer);
+
+    auto getColor = [](auto unpackedcolor) {
+        SoBaseColor* lineColor = new SoBaseColor;
+        float transparency;
+        SbColor lineCol(0.7f, 0.7f, 0.7f);
+        lineCol.setPackedValue(unpackedcolor, transparency);
+        lineColor->rgb.setValue(lineCol);
+        return lineColor;
+    };
+
+    //First we create the subdivision lines
+    createGridPart(GridNumberSubdivision, true,
+                   (GridNumberSubdivision == 1), GridLinePattern,
+                   getColor(GridLineColor), GridLineWidth);
+
+    //Second we create the wider lines marking every nth lines
+    if (GridNumberSubdivision > 1) {
+        createGridPart(GridNumberSubdivision, false, true,
+            GridDivLinePattern, getColor(GridDivLineColor), GridDivLineWidth);
+    }
+}
+
+void GridExtensionP::createGridPart(int numberSubdiv, bool subDivLines, bool divLines, int pattern, SoBaseColor* color, int lineWidth)
+{
+    SoGroup* parent = new Gui::SoSkipBoundingGroup();
+    GridRoot->addChild(parent);
+    SoVertexProperty* vts;
+
+    parent->addChild(color);
+
+    SoDrawStyle* DefaultStyle = new SoDrawStyle;
+    DefaultStyle->lineWidth = lineWidth;
+    DefaultStyle->linePattern = pattern;
+    parent->addChild(DefaultStyle);
+
+    SoPickStyle* PickStyle = new SoPickStyle;
+    PickStyle->style = SoPickStyle::UNPICKABLE;
+    parent->addChild(PickStyle);
+
+    SoLineSet* grid = new SoLineSet;
+    vts = new SoVertexProperty;
+    grid->vertexProperty = vts;
+
+    float gridDimension = 1.5 * camMaxDimension;
+    int vlines = static_cast<int>(gridDimension / computedGridValue); // total number of vertical lines
+    int nlines = 2 * vlines; // total number of lines
+
+    if (nlines > 2000) {
+        if(!isTooManySegmentsNotified) {
+            Base::Console().Warning("The grid is too dense, so it is being disabled. Consider zooming in or changing the grid configuration\n");
+            isTooManySegmentsNotified = true;
+        }
+
+        Gui::coinRemoveAllChildren(GridRoot);
+        return;
+    }
+    else {
+        isTooManySegmentsNotified = false;
+    }
+
+    // set the grid indices
+    grid->numVertices.setNum(nlines);
+    auto * vertices = grid->numVertices.startEditing();
+    for (int i = 0; i < nlines; i++)
+        vertices[i] = 2;
+    grid->numVertices.finishEditing();
+
+    // set the grid coordinates
+    vts->vertex.setNum(2 * nlines);
+    SbVec3f* vertex_coords = vts->vertex.startEditing();
+
+    float minX, minY, maxX, maxY, z;
+    camCenterOnSketch.getValue(minX, minY, z);
```

Comment:
This PR has a major conceptual flaw. It does not take into account the plane of projection.

CamCenterOnSketch (which should not even be called like that for an Object in Part, which knows nothing about the sketcher), is the center of the camera expressed in global 3D x,y,z. This means that the projected X,Y where the sketcher would be drawn, are not necessarily the X,Y stored in camCenterOnSketch, it may be that the camera is pointing towards the Y axis, and then global (X, Z) are the X,Y of the projected sketch. 

The consequence is that the grid only works reasonably well as the the camera focus (distance to the origin, zoom level) value is larger than the extent necessary for the grid in the projected sketch (X, Y).

The concept needs to be rethought.

Reply:
OK, can confirm. Even with slightly skewed view, as soon as going away from origin the grid doesn't display...

Reply:
With sufficient zoom-in, even with an unskewed full perpendicular view, the grid is chopped. The direction of chopping depends on the plane of attachment.

The grid needs to know the origin and direction of the plane on which it is supposed to draw the grid. Then it needs to retrieve the direction/position/focal distance of the camera from the viewer, to determine the amount of lines necessary to cover the area...